design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,159 @@
+
+* UNIX_DATE
+* DATETOSTRING
+* STRINGTODATE
+
+### Casting/Converting
+
+Since the above UDFs already handle conversions, casting between TIMESTAMP and STRING or BIGINT is
+unnecessary. If DATE and TIME types are implemented, then the conversions between them and TIMESTAMP
+will be as follows:
+
+| From      | To        | Result                                                |
+|---------- |:----------|:------------------------------------------------------|
+| TIMESTAMP | DATE      | DATE type representing date portion of the TIMESTAMP. |
+| TIMESTAMP | TIME      | TIME type representing time portion of the TIMESTAMP. |
+| DATE      | TIMESTAMP | TIMESTAMP with time portion set to 00:00:00.          |
+| TIME      | TIMESTAMP | TIMESTAMP with date portion set to Unix epoch.        |
+
+### Arithmetic operations and comparisons
+
+At the very least, we would want KSQL to be able to compare, add and subtract TIMESTAMPS.
+
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and [MariaDB](https://mariadb.com/kb/en/date-time-functions/)
+have very complete lists of built-in arithmetic functions, while [PostgreSQL](https://www.postgresql.org/docs/9.0/functions-datetime.html)
+defines behaviors for arithmetic operators.
+
+Because functions are more well-defined, we should opt for built-in arithmetic functions
+instead of using operators. The following functions are necessary:
+
+* ADDTIME(timestamp1, timestamp2)
+* SUBTIME(timestamp1, timestamp2)
+
+As for comparisons, the following expressions should be supported:
+```
+time_stamp1 < time_stamp2
+time_stamp1 > time_stamp2
+time_stamp1 = time_stamp2
+time_stamp1 BETWEEN time_stamp2 AND time_stamp3
+```
+
+Comparisons between TIMESTAMPS and other data types should not be allowed.
+
+### Window units
+
+It might make sense to give window units a timestamp representation so that they could be used as time intervals in TIMESTAMP arithmetic. For example,
+
+```roomsql
+SELECT ADDTIME(TIME, 1 DAY) FROM FOO;
+```
+
+Because windowing is handled separately from expression evaluation, doing this will not have any
+impact on windows.
+
+## Test plan
+
+There should be an integration test with Kafka Connect and Schema Registry.
+
+## LOEs and Delivery Milestones
+
+This feature can be broken down into two milestones: implementing the TIMESTAMP type (1-2 weeks) and
+adding all of the supporting UDFs (~1 week).
+
+## Documentation Updates
+
+There will need to be documentation on the following:
+
+* Description of the TIMESTAMP data type
+* TIMESTAMP usage in WHERE/GROUP/PARTITION clauses
+* New and updated UDFs
+* We might want to add this into one of the quick-starts
+
+## Compatibility Implications
+
+If a user creates a stream or table with a TIMESTAMP typed column, then that column will not be
+(de)serializable in earlier versions. Besides that, there should be no other compatibility issues. 
```

Comment:
Does it mean a value written as TIMESTAMP in 0.15 won't be deserialized in 0.14? Why not? If timestamps will be stored as long values in Delimited, Json, and Avro too; why will they fail?

Reply:
The situation I had in mind was if there was a command like `CREATE STREAM stream_name (col TIMESTAMP) AS ...`. KSQL would not recognize `TIMESTAMP` as a type. That was worded very inaccurately, I'll rewrite that part.

Timestamps in Avro/JSON etc will be stored and deserialized as long values in older versions.